webapp/src/ts/training-card.guard.provider.ts

```diff
@@ -0,0 +1,47 @@
+import { ActivatedRouteSnapshot, CanDeactivate, RouterStateSnapshot } from '@angular/router';
+import { Injectable } from '@angular/core';
+import { Store } from '@ngrx/store';
+
+import { GlobalActions } from '@mm-actions/global';
+import { Selectors } from '@mm-selectors/index';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TrainingCardDeactivationGuardProvider implements CanDeactivate<any> {
+  private globalActions;
+  private trainingCardOpen = false;
+  private formId: string | null = null;
+
+  constructor(private store: Store) {
+    this.globalActions = new GlobalActions(store);
+  }
+
+  canDeactivate(
+    component: any,
+    currentRoute: ActivatedRouteSnapshot,
+    currentState: RouterStateSnapshot,
+    nextState: RouterStateSnapshot,
+  ) {
+
+    this.store
```

Comment:
This subscription is asynchronous code. It can get slightly delayed and execute the `  if (this.trainingCardOpen) {` code block below first with old values in `this.trainingCardOpen` and `this.formId = trainingCard.formId;`

How can we make it safer?

Reply:
How about updating canDeactivate to return an Observable<boolean> and use `take` to wait for the store's latest value before executing `if (this.trainingCardOpen)` ?